src/main/org/deidentifier/arx/distributed/ARXPartition.java

```diff
@@ -69,164 +100,275 @@ public static Data getData(List<DataHandle> handles) {
         }
         return Data.create(new CombinedIterator<String[]>(iterators));
     }
+
     /**
-     * Partitions the dataset making sure that records from one equivalence
-     * class are assigned to exactly one partition. Will also remove all hierarchies.
-     * @param data
-     * @param number
-     * @return
+     * Splits a given dataset into a specified number of partitions, ensuring that each partition
+     * is within the same equivalence class based on quasi-identifying attributes. Each partition will have
+     * the same header as the original file.
+     *
+     * @param data           The dataset to be partitioned. (Expected to have a header)
+     * @param subsetIndices  A sorted array of indices specifying a subset of the data for partitioning.
+     * @param number         The number of partitions to create.
+     * @return               A list of {@link ARXPartition} objects representing the divided partitions.
      */
-    public static List<DataHandle> getPartitionsByClass(Data data, int number) {
+    public static List<ARXPartition> getPartitionsByClass(Data data, int[] subsetIndices, int number) {
+        DataHandle handle = data.getHandle();
 
-        // Prepare
-        List<DataHandle> result = new ArrayList<>();
         DataDefinition definition = data.getDefinition().clone();
-        DataHandle handle = data.getHandle();
         Set<String> qi = handle.getDefinition().getQuasiIdentifyingAttributes();
         for (String attribute : qi) {
             definition.resetHierarchy(attribute);
         }
-        
-        // Collect indices to use for sorting
-        int[] indices = new int[qi.size()];
-        int num = 0;
-        for (int column = 0; column < handle.getNumColumns(); column++) {
-            if (qi.contains(handle.getAttributeName(column))) {
-                indices[num++] = column;
-            }
-        }
-        
-        // Sort
-        handle.sort(true, indices);
-        
+
+        int[] qiIndices = createIndexArray(handle, qi);
+
+        // Shuffle
+        List<String[]> markedRows = sortData(handle, subsetIndices, qiIndices);
```

Comment:
Why is this called markedRows and not sortedRows?

Reply:
OK, I just realized that the whole way in which the sorting is implemented is a bit ugly. I suggest to not use "marks" to sort both the data and the indices at the same time. To achieve this, a common concept is to use a "Swapper". There are sorting methods, that allow you to provide a Comparator (for comparing elements) and a Swapper (for then swapping elements). In your case, you want to use a comparator as you did and a swapper, that not only swaps the data, but also swaps the indices in the subset.

Coming back to my suggestion above, it might be a good idea to use "RowSet" to represent the subset from the beginning, because RowSet offers a "swap" method.

Then you can use the CERN Colt sorting functions, which are already used at many places in ARX to do something like:

Sorting.mergeSort(comparator, swapper);

As a consequence, no functions are needed for marking anything and transforming back between data and incides.